content/posts/2021-11-13---Typy.md

```diff
@@ -0,0 +1,292 @@
+```
+public interface Code {
+    public void learnFromYoutube();
+    public int hoursToLearn();
+    public boolean isMyProjectFinnished();
+}
+```
+Jak widzimy, interfejs zawiera jedynie metody, które programista korzystający z danego interfejsu MUSI zaimplementować. Jest to w dużym skrócie po prostu gotowy szablon klasy, z której zamierzamy korzystać.
+Jest to niezwykle przydatne w praktyce, podczas planowania schematu naszego programu. Istotne jest to, że interfejsu mogą rozszerzać jedynie inne interfejsy. 
+Poza zwykłymi metodami, interfejsy mogą zawierać dodatkowo metody domyślne, prywatne, statyczne oraz stałe.
+Implementacja danego interfejsu jest również niezwykle prosta. Jedyne co należy zrobić, to zaimplementować metody tego interfejsu jak na poniższym przykładzie:
+```
+    Code code = new Code() {
+        @Override
+        public void learnFromYoutube() {
+
+        }
+
+        @Override
+        public int hoursToLearn() {
+            return 0;
+        }
+
+        @Override
+        public boolean isMyProjectFinnished() {
+            return false;
+        }
+    };
+```
+Tutaj widzimy domyślnie wygenerowane przez interfejs metody do implementacji. Adnotacja override sygnalizuje nam, że musimy te metody nadpisać.
+Poniżej można zapoznać się z przykładową implementacją tych metod:
+```
+    Code code = new Code() {
+        @Override
+        public void learnFromYoutube() {
+            System.out.println("Muszę obejrzeć w końcu ten kurs Javy...");
+        }
+
+        @Override
+        public int hoursToLearn() {
+            System.out.println("Będę się uczył dopóki nie osiągne pułapu 30 godzin.");
+            return 30;
+        }
+
+        @Override
+        public boolean isMyProjectFinnished() {
+            System.out.println("Nie no dopiero co zaczynam :(");
+            return false;
+        }
+    };
+```
+Przydatność tego rozwiązania możecie ocenić już sami, po przeczytaniu tej części wpisu.
+
+### Enumy
+Enumy to typ dodany w Javie 7. Jest to najprościej mówiąc zwykły typ wyliczeniowy, który pozwala programistom z góry zadeklarować konkretną liczbę możliwych wartości. 
+Tworzenie:
+```
+public enum Languages {
+    JAVA, PYTHON, C, PHP, JAVASCRIPT;
+}
+```
+W tym konkretnym przypadku skorzystaliśmy z typu wyliczeniowego do deklaracji określonej liczby języków programowania. Konwencja nazwnictwa tego co wyliczamy, widoczna jest powyżej.
+Pisanie małą literą jest dozwolone, ale tak jak wyżej już zwracałem uwagę - powinno się pisać z wielkiej litery. 
+Enumy można również opakować w klasę, aby potem uzyskać szerszą funkcjonalność tego typu:
+```
+public class Languages{
+    public enum NaszEnum {
+        JAVA, PYTHON, C, PHP, JAVASCRIPT;
+    }
+}
+```
+Zastosowanie enuma w praktyce, polega na utworzeniu obiektu klasy (w tym wypadku) Languages i skorzystania w ramach niego z naszego enuma, tak jak na poniższym przykładzie:
```

Comment:
Pokaż zastosowanie enuma bez takiego zagnieżdzenia, bo to nie wygląda najlepiej, np. daj że Language ma pole bedące enumem - jest to typowe zastosowanie ich